src/main/scala/com/mozilla/telemetry/learning/federated/FederatedLearningSearchOptimizerS3Sink.scala

```diff
@@ -0,0 +1,151 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package com.mozilla.telemetry.learning.federated
+
+import com.mozilla.telemetry.learning.federated.FederatedLearningSearchOptimizerConstants.{NumberOfFeatures, StartingLearningRate, StartingWeights}
+import com.mozilla.telemetry.streaming.FrecencyUpdateAggregate
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.fs.Path
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.sql.sources.StreamSinkProvider
+import org.apache.spark.sql.streaming.OutputMode
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.json4s.DefaultFormats
+import org.json4s.jackson.Serialization
+
+object FederatedLearningSearchOptimizerConstants {
+  // https://github.com/florian/federated-learning-addon/blob/master/addon/prefs.js#L26
+  val StartingWeights: Array[Double] = Array(4, 14, 31, 90, 100, 70, 50, 30, 10, 0, 0, 100, 2000, 75, 0, 0, 0, 25, 0, 140, 200, 0)
+  val NumberOfFeatures: Int = StartingWeights.length
+  val StartingLearningRate: Int = 2
+}
+
+class FederatedLearningSearchOptimizerS3Sink(outputPath: String, stateCheckpointPath: String) extends Sink {
+
+  private[federated] var state: OptimisationState = initState()
+
+  override def addBatch(batchId: Long, data: DataFrame): Unit = {
+    val aggregates = data.collect().map(FrecencyUpdateAggregate(_))
+
+    val iteration: Long = state.iteration
+
+    val ord = Ordering.by((_: FrecencyUpdateAggregate).modelVersion)
+    aggregates.filter(_.studyVariation == "treatment").filter(_.modelVersion.toLong >= iteration).reduceOption(ord.min) match {
+      case None =>
+      // no relevant updates, do nothing
+      case Some(aggregate) =>
+
+        val lastWeights: Array[Double] = state.weights
+        val learningRates: Array[Double] = state.learningRates
+        val previousGradient: Option[Array[Double]] = state.gradient
+
+        val gradient = aggregate.avgUpdates
+        val InternalResult(newWeights, newLearningRates) = FederatedLearningRPropOptimizer.fit(lastWeights, gradient, previousGradient, learningRates)
+
+        val newIteration = iteration + 1
+
+        // val log = Array(OptimisationLogLine(aggregate.window.start, aggregate.modelVersion.toLong, aggregate.avgLoss))
+
+        val newState = OptimisationState(newIteration, newWeights, newLearningRates, Option(gradient))
+        writeModel(ModelOutput(newWeights.map(math.round(_).toInt), newIteration))
+
+        writeState(newState)
+        state = newState
+    }
+  }
+
+  private[federated] def writeModel(modelOutput: ModelOutput): Unit = {
```

Comment:
I'm not sure if you're already giving the file to s3 here, but two points about s3:
- The file permissions need to be changed so that the file is accessible for everyone
- Caching should be disabled (HTTP header `Cache-Control: no-cache` so that clients always fetch the newest model)

Anthony did both of these for me so that I could test my addon, but he needed to explicitly configure both of them.

Reply:
It seems likely we'll have to switch to an S3-specific library for these requirements